File: src/系统/03．技能系统/00．技能模板+函数/04．机制组件/09．装备通用机制/16．单位时限数值.lua
local ____lualib = require("lualib_bundle")
local __TS__Delete = ____lualib.__TS__Delete
local ____exports = {}
---
-- @noSelfInFile
local jass = require("jass.common")
local ____require_result_0 = require("系统.00．核心系统.05．中心计时器")
local getServerTime = ____require_result_0.getServerTime
local GetHandleId = jass.GetHandleId
local function _____53D6_5355_4F4DID(unit)
    if unit == nil or unit == 0 then
        return 0
    end
    return GetHandleId(unit) or 0
end
____exports["创建单位时限数值"] = function(_____540D_79F0)
    local _____8868 = {}
    local function _____8BFB_53D6_8BB0_5F55(unit)
        local id = _____53D6_5355_4F4DID(unit)
        if id == 0 then
            return nil
        end
        local _____8BB0_5F55 = _____8868[id]
        if _____8BB0_5F55 == nil then
            return nil
        end
        if _____8BB0_5F55["到期"] >= getServerTime() then
            return _____8BB0_5F55
        end
        __TS__Delete(_____8868, id)
        return nil
    end
    return {
        ["名称"] = _____540D_79F0,
        ["写入"] = function(unit, _____503C, _____6301_7EED_79D2)
            local id = _____53D6_5355_4F4DID(unit)
            if id == 0 or not (_____6301_7EED_79D2 > 0) then
                return
            end
            _____8868[id] = {
                ["值"] = _____503C,
                ["到期"] = getServerTime() + _____6301_7EED_79D2 * 1000
            }
        end,
        ["读取"] = function(unit)
            local ____opt_1 = _____8BFB_53D6_8BB0_5F55(unit)
            return ____opt_1 and ____opt_1["值"]
        end,
        ["存在"] = function(unit)
            return _____8BFB_53D6_8BB0_5F55(unit) ~= nil
        end,
        ["消耗"] = function(unit)
            local id = _____53D6_5355_4F4DID(unit)
            if id == 0 then
                return nil
            end
            local _____8BB0_5F55 = _____8BFB_53D6_8BB0_5F55(unit)
            if _____8BB0_5F55 == nil then
                return nil
            end
            __TS__Delete(_____8868, id)
            return _____8BB0_5F55["值"]
        end,
        ["清空"] = function(unit)
            if unit == nil then
                _____8868 = {}
                return
            end
            __TS__Delete(
                _____8868,
                _____53D6_5355_4F4DID(unit)
            )
        end
    }
end
return ____exports

File: TS/系统/03．技能系统/00．技能模板+函数/04．机制组件/09．装备通用机制/16．单位时限数值.ts
/** @noSelfInFile */

const jass = require("jass.common") as any;
const { getServerTime } = require("系统.00．核心系统.05．中心计时器") as {
  getServerTime: (this: void) => number;
};

const GetHandleId = jass.GetHandleId as (handle: any) => number;

export interface 单位时限数值控制器 {
  readonly 名称: string;
  写入(this: void, unit: any, 值: number, 持续秒: number): void;
  读取(this: void, unit: any): number | undefined;
  存在(this: void, unit: any): boolean;
  消耗(this: void, unit: any): number | undefined;
  清空(this: void, unit?: any): void;
}

interface 单位时限数值记录 {
  值: number;
  到期: number;
}

function 取单位ID(this: void, unit: any): number {
  if (unit == null || unit === 0) return 0;
  return GetHandleId(unit) || 0;
}

export function 创建单位时限数值(this: void, 名称: string): 单位时限数值控制器 {
  let 表: Record<number, 单位时限数值记录 | undefined> = {};

  function 读取记录(this: void, unit: any): 单位时限数值记录 | undefined {
    const id = 取单位ID(unit);
    if (id === 0) return undefined;
    const 记录 = 表[id];
    if (记录 == null) return undefined;
    if (记录.到期 >= getServerTime()) return 记录;
    delete 表[id];
    return undefined;
  }

  return {
    名称,
    写入: function 写入(this: void, unit: any, 值: number, 持续秒: number): void {
      const id = 取单位ID(unit);
      if (id === 0 || !(持续秒 > 0)) return;
      表[id] = { 值, 到期: getServerTime() + 持续秒 * 1000 };
    },
    读取: function 读取(this: void, unit: any): number | undefined {
      return 读取记录(unit)?.值;
    },
    存在: function 存在(this: void, unit: any): boolean {
      return 读取记录(unit) != null;
    },
    消耗: function 消耗(this: void, unit: any): number | undefined {
      const id = 取单位ID(unit);
      if (id === 0) return undefined;
      const 记录 = 读取记录(unit);
      if (记录 == null) return undefined;
      delete 表[id];
      return 记录.值;
    },
    清空: function 清空(this: void, unit?: any): void {
      if (unit == null) {
        表 = {};
        return;
      }
      delete 表[取单位ID(unit)];
    },
  };
}

export {};
